Crash when QNetworkReply is aborted due to timeout

I'm in China and can't access YouTube without a VPN.
There is a crash problem here.
The call to setTransferTimeout() is bound to crash if it doesn't reply with data within the time set by the parameter.

``` c++
void MainWindow::on_pushButton_clicked()
{
    fetchData(QString("https://www.youtube.com/"))
    .then([](const QByteArray &bytes) {
        qDebug().noquote() << "fetch data success:" << bytes;
    }, [](const std::exception& e) {
        qDebug() << "fetch data error:" << e.what();
    });
}

QCoro::Task<QByteArray> MainWindow::fetchData(const QUrl &url)
{
    qDebug() << "request:" << url.toString();
    QNetworkRequest request(url);
    // request.setTransferTimeout(10000);
    QNetworkAccessManager manager;
    const auto reply = co_await manager.get(request);
    if (reply->error() != QNetworkReply::NoError) {
        qDebug() << "request error" << reply->errorString();
        co_return {};
    }
    co_return reply->readAll();
}
```

running environment: Qt_5_15_2_MSVC2019_32bit-Debug

![image](https://github.com/user-attachments/assets/bea4b764-e1c3-4a98-ac5a-6d83e4a54a9c)
C:\Users\qt\work\qt\qtbase\src\corelib\kernel\qobject.cpp:3770: error: Debugger encountered an exception: Exception at 0x7b066f0b, code: 0xc0000005: read access violation at: 0xfeeefefe, flags=0x0 (first chance)

--- a/qcoro/network/qcoronetworkreply.cpp
+++ b/qcoro/network/qcoronetworkreply.cpp
@@ -15,55 +15,69 @@ class ReplyWaitSignalHelper : public WaitSignalHelper {
 public:
     ReplyWaitSignalHelper(const QNetworkReply *reply, void(QIODevice::*signal)())
         : WaitSignalHelper(reply, signal)
-        #if QT_VERSION >= QT_VERSION_CHECK(5, 15, 0)
-        , mError(connect(reply, &QNetworkReply::errorOccurred, this, [this]() { emitReady(false); }))
-        #endif
-        , mFinished(connect(reply, &QNetworkReply::finished, this, [this]() { emitReady(true); }))
+        , mError(connect(reply, &QNetworkReply::errorOccurred, this, [this]() { emitReady(false); }, Qt::QueuedConnection))
+        , mFinished(connect(reply, &QNetworkReply::finished, this, [this]() { emitReady(true); }, Qt::QueuedConnection))
     {}
     ReplyWaitSignalHelper(const QNetworkReply *reply, void(QIODevice::*signal)(qint64))
         : WaitSignalHelper(reply, signal)
-        #if QT_VERSION >= QT_VERSION_CHECK(5, 15, 0)
-        , mError(connect(reply, &QNetworkReply::errorOccurred, this, [this]() { emitReady(0LL); }))
-        #endif
-        , mFinished(connect(reply, &QNetworkReply::finished, this, [this]() { emitReady(0LL); }))
+        , mError(connect(reply, &QNetworkReply::errorOccurred, this, [this]() { emitReady(0LL); }, Qt::QueuedConnection))
+        , mFinished(connect(reply, &QNetworkReply::finished, this, [this]() { emitReady(0LL); }, Qt::QueuedConnection))
     {}
 
 private:
     void cleanup() override {
-        #if QT_VERSION >= QT_VERSION_CHECK(5, 15, 0)
         disconnect(mError);
-        #endif
         disconnect(mFinished);
         WaitSignalHelper::cleanup();
     }
 
-    #if QT_VERSION >= QT_VERSION_CHECK(5, 15, 0)
     QMetaObject::Connection mError;
-    #endif
     QMetaObject::Connection mFinished;
 };
 
 } // namespace
 
+struct QCoroNetworkReply::WaitForFinishedOperation::Private {
+    static_assert(sizeof(std::unique_ptr<Private>)  + sizeof(void*) == sizeof(QPointer<QNetworkReply>),
+                  "QCoroNetworkReply::WaitForFinishedOperation is not BC with previous version, see comment in header");
+
+    Private(QPointer<QNetworkReply> reply)
+        : reply(reply)
+    {
+        // Ensure the signal is emitted in the same thread as the reply, not on the thread
+        // where the coroutine is resumed...
+        if (reply) {
+            dummy.moveToThread(reply->thread());
+        }
+    }
+
+    QPointer<QNetworkReply> reply;
+    QObject dummy;
+};
+
 QCoroNetworkReply::WaitForFinishedOperation::WaitForFinishedOperation(QPointer<QNetworkReply> reply)
-    : mReply(reply)
-{}
+    : d(std::make_unique<Private>(reply))
+{
+}
+
+QCoroNetworkReply::WaitForFinishedOperation::~WaitForFinishedOperation() = default;
 
 bool QCoroNetworkReply::WaitForFinishedOperation::await_ready() const noexcept {
-    return !mReply || mReply->isFinished();
+    return !d->reply || d->reply->isFinished();
 }
 
 void QCoroNetworkReply::WaitForFinishedOperation::await_suspend(std::coroutine_handle<> awaitingCoroutine) {
-    if (mReply) {
-        QObject::connect(mReply, &QNetworkReply::finished,
-                         [awaitingCoroutine]() mutable { awaitingCoroutine.resume(); });
+    if (d->reply) {
+        QObject::connect(d->reply, &QNetworkReply::finished, &d->dummy,
+                         [awaitingCoroutine]() mutable { awaitingCoroutine.resume(); },
+                         Qt::QueuedConnection);
     } else {
         awaitingCoroutine.resume();
     }
 }
 
 QNetworkReply *QCoroNetworkReply::WaitForFinishedOperation::await_resume() const noexcept {
-    return mReply;
+    return d->reply;
 }
 
 QCoro::Task<std::optional<bool>> QCoroNetworkReply::waitForReadyReadImpl(std::chrono::milliseconds timeout) {

--- a/qcoro/network/qcoronetworkreply.h
+++ b/qcoro/network/qcoronetworkreply.h
@@ -16,13 +16,24 @@ private:
     class WaitForFinishedOperation final {
     public:
         explicit WaitForFinishedOperation(QPointer<QNetworkReply> reply);
+        ~WaitForFinishedOperation();
 
         bool await_ready() const noexcept;
         void await_suspend(std::coroutine_handle<> awaitingCoroutine);
         QNetworkReply *await_resume() const noexcept;
 
     private:
-        QPointer<QNetworkReply> mReply;
+        struct Private;
+        std::unique_ptr<Private> d;
+        // BC: This class used to have a QPointer<QNetworkReply>, which is 2*sizeof(void*),
+        // while std::unique_ptr is only sizeof(void*), so this dummy one is to ensure binary
+        // compatibility of this class.
+        // FIXME: Remove in 1.0
+        // Silly gcc 11, cannot detect the dummy is unused and warns about unused attribute
+        #if __GNUC__ > 11 || defined(__clang__) || defined(_MSC_VER)
+        [[maybe_unused]]
+        #endif
+        void *dummy = nullptr;
     };
 
     friend struct awaiter_type<QNetworkReply *>;

--- a/tests/qcoronetworkreply.cpp
+++ b/tests/qcoronetworkreply.cpp
@@ -167,6 +167,19 @@ private:
         QVERIFY(called);
     }
 
+    // See https://github.com/danvratil/qcoro/issues/231
+    QCoro::Task<> testAbortOnTimeout_coro(QCoro::TestContext) {
+        auto request = buildRequest(QStringLiteral("block"));
+        request.setTransferTimeout(300);
+        QNetworkAccessManager nam;
+        auto reply = co_await nam.get(request);
+        QCORO_VERIFY(reply != nullptr);
+        QCORO_VERIFY(reply->isFinished());
+        QCORO_COMPARE(reply->error(), QNetworkReply::OperationCanceledError);
+        // QNAM is destroyed here and so is all its associated state, which could
+        // crash (or cause invalid memory access)
+    }
+
 private Q_SLOTS:
     void init() {
         mServer.start(QHostAddress::LocalHost);
@@ -184,6 +197,7 @@ private Q_SLOTS:
     addCoroAndThenTests(ReadAllTriggers)
     addCoroAndThenTests(ReadTriggers)
     addCoroAndThenTests(ReadLineTriggers)
+    addTest(AbortOnTimeout)
 
 private:
     QNetworkRequest buildRequest(const QString &path = QString()) {

